### PR TITLE
Added installer for Docker 20.10.8, 20.10.9, 20.10.10 and 20.10.11

### PR DESCRIPTION
**Pull request by rayandas**
https://github.com/rancher/rancher/issues/33926
https://github.com/rancher/rancher/issues/35031
https://github.com/rancher/rancher/issues/35272
- This PR adds installer for Docker `20.10.8`, `20.10.9` and `20.10.10`
- Excludes Docker `20.10.8` and higher versions from testing on Ubuntu 16.04 (Docker 20.10.8 install fails on it) because Docker has removed `20.10.8` packaging on Ubuntu 16.04 as it reached EOL as per https://docs.docker.com/engine/release-notes/#packaging

### CODE REVIEW

**Review by paynejacob — DISMISSED** [review]

File: pkg/20.10.8/config.sh
Comment (rawmind0) [inline pkg/20.10.8/config.sh:8]: If this docker version shouldn't be tested at ubuntu `16.04`, it shouldn't be added here.

```
-TEST_OS_IMAGE_TAG[2]="16.04 18.04 20.04"
+TEST_OS_IMAGE_TAG[2]="18.04 20.04"
```

**Review by rawmind0 — CHANGES_REQUESTED** [review]

File: pkg/20.10.8/config.sh
Comment (rayandas) [reply pkg/20.10.8/config.sh:8]: @rawmind0 Ah! right. let me change it right away.

**Review by rawmind0 — DISMISSED** [review]
LGTM

**Comment by deniseschannon** [comment]
As of now, this PR needs to be on hold until we get clarity from PM on how we want to support different versions. Merging this will break SLES installs using our Docker installs.

**Comment by rayandas** [comment]
@deniseschannon Fixed the script. Now if we use `curl https://releases.rancher.com/install-docker/20.10.sh | sh` this will run `20.10.8` Docker install script. If Docker `20.10.8` package is not found in an OS, except breaking from the script, it will lower the version to `20.10.7` and try to install again, if it doesn't find `20.10.7` then it'll lower the version again and try to install `20.10.6` and so on. 
After every retry it'll check if `which docker` returns true or not. If Docker gets installed, it will return true and break from the loop.

**Comment by rayandas** [comment]
TODO: Add docker install script `20.10.9` and `20.10.10` on this PR and update it.

**Review by superseb — CHANGES_REQUESTED** [review]
Please add the design outcome or link to the design page as there is no way to validate what technically is implemented is according to design. My questions based on this PR, what is the scope of "try to install a lower version"? Are we accepting that this is a new way of doing things in only a few patch versions and also that it no longer aligns with having script names with the exact patch or minor version giving you the expected result? How are we communicating to users that you now no longer get the version you are asking for?

**Comment by rayandas** [comment]
@superseb I have updated the PR description with the workflow (how it's working). @deniseschannon and our team had a discussion around it and we decided to modify the script and make it intelligent enough so that the script automatically try to install the next lower version available for the OS from https://github.com/rancher/install-docker/tree/master/dist (20.10.10 -> 20.10.9 -> 20.10.8 and so on.) 
Also @deniseschannon has a plan to release a Docker support matrix, where user can find which latest docker version works on particular OSes using node driver. Talks are on-going.
Please let me know if you need anything.

**Comment by vincent99** [comment]
This retry lower logic shouldn't be in _every_ version file.  These scripts need to give you what they say they're going to.

  - If I ask for x.y.z.sh, I should get exactly x.y.z, or an error.
  - If I ask for x.y.sh then I don't care about the patch version, so you can try z, z-1,... 0 to find one that works.  But shouldn't go beyond that and start changing x or y.  These files don't _need_ to be symlinks, that was just convenient before.  This is where your retrying logic should be.
  - If you want to cover going farther beyond that and finding any version that works, make another and call it like latest.sh.

**Comment by deniseschannon** [comment]
I agree with what @vincent99 is saying. I thought we had discussed that the lowering version logic would be in the x.y.sh files and not the individual x.y.z.sh files due to how Rancher calls the Docker install script.

**Comment by rayandas** [comment]
@vincent99 @deniseschannon I got the point. Will update the PR with all the suggestions.

**Comment by rayandas** [comment]
Hey all, I updated this PR. I removed everything related to the changes where we were trying to make the script a bit intelligent. I saw there will be quite a few changes that needs to be done (dist/ pkg/ tests, README and I have those changes on another branch). I thought I will create a separate PR for that. It'll be easy for reviewers also. 

So this PR contains only the normal scripts for `20.10.8`, `20.10.9` and `20.10.10` Docker versions. (the general procedure to add a new installer for docker.)
@deniseschannon @superseb @samkulkarni20

**Comment by rayandas** [comment]
This PR is ready again for review. There is another PR for the independent script.  #89

**Comment by snasovich** [comment]
@superseb , could you please take another look as you're showing as requesting changes and I want to make sure your comments are addressed.

**Comment by superseb** [comment]
@snasovich The questions were around the design and if that matched the implementation which was unclear. This PR is now simplified to adding new versions as-is. 20.10.11 is also available at the moment.

**Comment by rayandas** [comment]
@superseb should I add 20.10.11 version too in this PR?

**Comment by superseb** [comment]
Yes please

**Comment by rayandas** [comment]
Added v20.10.11 and removed 20.10.sh symlink with latest versions because we will add an independent script as 20.10.sh (which will install the latest available docker version automatically for the given OS.) in another PR, see https://github.com/rancher/install-docker/pull/89.

**Review by superseb — APPROVED** [review]

**Comment by snasovich** [comment]
@paynejacob , could you please review it on behalf of our team, please?

**Comment by rayandas** [comment]
Before anyone merge it, I want to point out one thing again, in this PR the symlink `20.10.sh` (which was symlinked with the latest docker version before.) is removed. 
`20.10.sh` has been added as an independent script which will install the latest available docker version automatically for the given OS, this is in another PR #89 

cc: @superseb @paynejacob

**Review by paynejacob — APPROVED** [review]

**Pull request merged**
